### PR TITLE
Tutorial 4 - Lattice-Boltzmann

### LINKED ISSUE #3828: Tutorial 04: lb part1
* It should be stated at the top that LB is for hydrodynamics

* Poiseuille  flow needs to be described more accurately than "flow between tow walls".

* LB_BOUNDARIES should be listed as recommended myconfig feature, so it also works with the CPU.

* The info about there being four relaxation rates is repeated immediately after the 1st statement

* kT needs to be added to the list of LbFluid parameters in the text

* The statement that LB Boundaries is an experimental feature needs to be removed

* boundary is initialized -> instanced

### PR DESCRIPTION
**Pull request by jngrad**
Description of changes:
- Update part 1 (fixes #3828)
- Write solutions for parts 2 and 3 and check them in CI
- Calculate the diffusion coefficient of a polymer with the MSD and Green-Kubo methods
- Replace LB hydrodynamics by LD hydrodynamics for polymer diffusion

### CODE REVIEW

File: doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb [hunk 1]
```diff
@@ -213,7 +213,7 @@
     "<figure>\n",
     "<img src='figures/latticeboltzmann-momentumexchange.png', style=\"width: 300px;\"/>\n",
     "<center>\n",
-    "<figcaption>The coupling scheme between fluid and particles is based on the interpolation of the fluid velocity $\\vec{u}$ from the grid nodes. This is done by trilinear interpolation. The difference between the particle velocity $\\vec{v}(t)$ and the interpolated velocity $\\vec{u}(\\vec{r},t)$ is used in the momentum exchange of the equation $\\vec{F}_H$ above.</figcaption>\n",
```
Comment (KaiSzuttor) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb]: it is actually trilinear... https://en.wikipedia.org/wiki/Trilinear_interpolation

Comment (jngrad) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb]: thanks!

**Comment by jngrad** [comment]
@KaiSzuttor Part 3 is now ready for review. The idea would be to hide the solution of the plots for Rf, Rg, Rh. Once the participants have a working solution for Rf, they can copy-paste it for Rg and Rh. Regarding the diffusion coefficient, they should re-use the solution they wrote in Part 2 and write a fitting function based on the Kirkwood-Zimm model.

I'm still working on the electrophoresis part, but if I can't get it to work out in time, plan B is to only do the monomer+polymer diffusion in the advanced group, and the monomer + Poiseuille flow in the beginner group.

I tried Part 3 with Langevin Dynamics instead of LB, and I got similar results at a fraction of the computational time. Maybe in the future we could create a polymer tutorial with Langevin.

**Comment by jngrad** [comment]
@schlaicha @christophlohrmann @reinaual the latest changes are in, I'll hide solutions later

**Comment by RudolfWeeber** [comment]
I only had a quick look so far.
Here are my first impressions, which might not be accurate.

One thing I find very important is to make it very clear in the text, what part of the observations actually depend on hydrodynamics and which don't.

Stuff like r_g shouldn't really be calculated with active LB, since the results can only get worse due to longer simulation times.
Also for the single particle diffusion, what difference does the LB make?

Long term, it might even be reasonable to split this tutorial. The equilibrium polymer properties could be handled in a separate "Polymers"tutorial which wouldn't use LB.

**Comment by jngrad** [comment]
@RudolfWeeber I get the same results with Langevin Dynamics in place of LB for all observables in the polymer part. The same is probably true for the monomer part, although I didn't try it.

File: doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb [hunk 2]
```diff
@@ -9,11 +9,12 @@
     "#### Before you start:\n",
```
Comment (schlaicha) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <blockquote>After almost 20 years of developmentI</blockquote><p>I disagree, in 2020 we have more than 30 years of development: cellular automata were popular in the 1980s and Succi's first paper with ther term "Lattice Boltzmann" is from 1989, not sure if this was the first paper which introduced the name though.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/3893/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (schlaicha) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <p>I guess the equation should read </p><pre class="ql-syntax" spellcheck="false">\begin{align}
&nbsp;\rho &amp;= \sum_i f_i \
&nbsp;\vec{j} = \rho \\
&nbsp;\vec{u} &amp;= \sum_i f_i \vec{c_i} \\
&nbsp;\Pi^{\alpha \beta} &amp;= \sum_i f_i \vec{c_i}^{\alpha}\vec{c_i}^{\beta}
&nbsp;\label{eq:fields}
\end{align}
</pre><p>for better readability?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/3893/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (schlaicha) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <p>Again here the equations don't get typeset correctly in the notebook? Or is this due to the HTML export we perform later?</p><p> </p><ul><li>add a note that thermalization is optional</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/3893/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (schlaicha) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <ul><li>add links to the online documentation here to be consistent with the other tutorials</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/3893/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (schlaicha) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <ul><li>minimum example -&gt; minimal example</li><li> Initialize and LBFluid -&gt; no "and"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/3893/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (schlaicha) [inline doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <ul><li>Add links to the documentation here, too.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/3893/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

**Comment by schlaicha** [comment]
The numbering is a bit weird as part 1 is from 1-3, part 2 starts with 5 and part 3 with 5.2; splitting into several files seems like a good idea but should be introduced in the notebook of part 1.

**Comment by schlaicha** [comment]
> Long term, it might even be reasonable to split this tutorial. The equilibrium polymer properties could be handled in a separate "Polymers"tutorial which wouldn't use LB.

Very good point!

**Comment by christophlohrmann** [comment]
In Part 4 I think we should switch the exercise, such that the students have to set up the LB stuff and we give them the comparison to theory

File: doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb (hunk 2 above)
Comment (jngrad) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <p>good point, thanks!</p><p> </p>

Comment (jngrad) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <p>ReviewNB's support for LaTeX seems incomplete, the original equation looks good in Jupyter, with one equation per line</p>

Comment (jngrad) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <p>done</p>

Comment (jngrad) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: <p>done</p>

Comment (jngrad) [reply doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part1.ipynb:9]: done

**Comment by jngrad** [comment]
@schlaicha the numbering now restarts from 1 in every part
@christophlohrmann good idea, done

**Comment by RudolfWeeber** [comment]
I suppose, we merge this PR this morning, to avoid confusion?

**Comment by jngrad** [comment]
Yes, that would be more convenient for tutors.

**Review by RudolfWeeber — APPROVED** [review]

**Pull request merged**